### PR TITLE
Do probes in Go if Go-DOH is active

### PR DESCRIPTION
**Pull request by bemasc**

### CODE REVIEW

File: Android/app/src/main/java/app/intra/net/doh/Race.java [hunk 1]
```diff
@@ -33,38 +36,69 @@
...
+      probes.add(new GoProbe(context, urls[i], new Callback(i, probes, listener)));
     }
-    this.listener = listener;
+    return new Race(probes);
   }
 
-  public synchronized void start() {
-    for (Probe task : probes) {
-      task.start();
+  public static Race get(Context context, String[] urls, Listener listener) {
```
Comment (fortuna) [inline Android/app/src/main/java/app/intra/net/doh/Race.java]: Make the `start()` calls in this method too. No need to put it in a separate method. `Race` can be fully static.

File: Android/app/src/main/java/app/intra/net/doh/ServerConnectionProbe.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import androidx.annotation.NonNull;
+import app.intra.net.dns.DnsUdpQuery;
+import java.io.IOException;
+import okhttp3.Call;
+import okhttp3.Response;
+
+/**
+ * Implements a Probe using the OkHttp-based DoH client.
+ */
+class ServerConnectionProbe extends Probe {
```
[the remark this answers is not in the document]
Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/ServerConnectionProbe.java]: Done.  I went with `JavaProbe`.

File: Android/app/src/main/java/app/intra/net/doh/Race.java [hunk 3]
```diff
@@ -33,38 +36,69 @@
...
 
   /**
    * Creates a race between different servers.  To run the race, call start().
    * @param factory The factory will be called once for each server.
    * @param urls The URLs for all the DOH servers to compare.
    * @param listener Called once on an arbitrary thread with the result of the race.
    */
-  public Race(ServerConnectionFactory factory, String[] urls, Listener listener) {
-    probes = new ArrayList<>(urls.length);
+  static Race serverConnectionRace(ServerConnectionFactory factory,
```
[the remark this answers is not in the document]
Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/Race.java]: This code was removed in pursuit of your other comment.

File: Android/app/src/main/java/app/intra/net/doh/Race.java [hunk 4]
```diff
@@ -33,38 +36,69 @@
   }
 
   private final List<Probe> probes;
-  private final Listener listener;
+  private Race(List<Probe> probes) {
```
[the remark this answers is not in the document]
Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/Race.java]: Done.  This required using a factory pattern to maintain unit test access.

File: Android/app/src/main/java/app/intra/net/doh/Race.java [hunk 5]
```diff
@@ -33,38 +36,69 @@
...
+   * Creates a race between different servers using Go.  To run the race, call start().
+   * @param context Used to read the IP addresses of the servers from storage.
+   * @param urls The URLs for all the DOH servers to compare.
+   * @param listener Called once on an arbitrary thread with the result of the race.
+   */
+  private static Race goRace(Context context, String[] urls, Listener listener) {
+    List<Probe> probes = new ArrayList<>(urls.length);
     for (int i = 0; i < urls.length; ++i) {
-      probes.add(new Probe(factory, urls[i], new Callback(i)));
+      probes.add(new GoProbe(context, urls[i], new Callback(i, probes, listener)));
```
[the remark this answers is not in the document]
Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/Race.java]: Done

File: Android/app/src/main/java/app/intra/net/doh/Race.java (hunk 1 above)
Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/Race.java]: Done

File: Android/app/src/main/java/app/intra/net/doh/Race.java [hunk 6]
```diff
@@ -32,39 +35,45 @@
...
   }
 
-  public synchronized void start() {
-    for (Probe task : probes) {
-      task.start();
+  // Exposed for unit testing only.
+  static void start(Probe.Factory factory, Context context, String[] urls, Listener listener) {
+    List<Probe> probes = new ArrayList<>(urls.length);
+    for (int i = 0; i < urls.length; ++i) {
+      probes.add(factory.get(context, urls[i], new Callback(i, probes, listener)));
```
Comment (fortuna) [inline Android/app/src/main/java/app/intra/net/doh/Race.java]: It doesn't seem like you need a factory. You could simply pass a function that takes a URL and a done(error) callback and runs the test. No need to create an object for each URL.

The use of protected methods is usually a Code Health red flag. The functionality you are reusing via inheritance can be here instead, and the Probes can simply be functions. Or a simple class with a single test/run method as described above.

Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/Race.java]: Done.  This is much simpler.

This design no longer supports cancelling outstanding probes, but probes are not expensive so cancelling them is not very important, and in the previous implementation cancellation was only partially supported.  (Cancellation was only possible before the probe query is sent).

File: Android/app/src/main/java/app/intra/net/doh/JavaProber.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  @Override
+  public void probe(String url, Callback callback) {
+    new Thread(() -> {
+      // Factory.get() is a slow, synchronous call, so it has to run in a new thread.
+      ServerConnection conn = factory.get(url);
+      if (conn == null) {
+        callback.onCompleted(false);
+        return;
+      }
+      conn.performDnsRequest(QUERY, QUERY_DATA, new QueryCallback(callback));
```
Comment (fortuna) [inline Android/app/src/main/java/app/intra/net/doh/JavaProber.java]: It seems we can remove the DnsUdpQuery parameter from `performDnsRequest` and the `QUERY` variable here.

Comment (bemasc) [reply Android/app/src/main/java/app/intra/net/doh/JavaProber.java]: Done

**Pull request merged**